Ensure data are ready for comms libraries with events

wence- opened the pull request:
Rather than using device synchronisation, use events to track work on streams for buffers and only hand over the buffer data pointer to the comms library when the event is ready.

wence- commented:
An alternative approach (borrowing a bunch of the same ideas) to #225.

wence- commented on cpp/src/cuda_event.cpp:
```diff
@@ -0,0 +1,60 @@
+    }
+}
+
+void Event::add_as_dependency_to(rmm::cuda_stream_view stream) const {
+    if (!query()) {
+        RAPIDSMPF_CUDA_TRY(
+            cudaStreamWaitEvent(stream.value(), event_, cudaEventWaitDefault)
+        );
+    }
+}
```
This is preparing for if we start using an internal stream inside the shuffler that is distinct from the user stream. In that case, when we copy a user buffer on the internal stream we must add a dependency on the work on the user's stream.

wence- commented on cpp/src/shuffler/shuffler.cpp:
```diff
@@ -485,7 +502,10 @@ void Shuffler::insert(std::unordered_map<PartID, PackedData>&& chunks) {
+                pid,
+                std::move(packed_data.metadata),
+                std::move(packed_data.gpu_data),
+                event
```
Chunk creation takes an event, and forwards it to the buffer.

pentschev commented on cpp/include/rapidsmpf/buffer/buffer.hpp:
```diff
@@ -105,6 +111,28 @@ class Buffer {
      */
     [[nodiscard]] void const* data() const;
 
+    /**
+     * @brief ensure all stream-ordered work to populate the buffer is completed.
+     */
+    void synchronize() const;
```
Ok, but how do we prevent abuse then? In my mind, providing an interface like this is recipe to "go do the easy thing".

wence- commented on cpp/src/buffer/resource.cpp:
```diff
@@ -88,15 +90,20 @@ std::unique_ptr<Buffer> BufferResource::allocate(
+                std::unique_ptr<Buffer>(new Buffer{std::move(buf), this, std::move(event)}
+                );
+            break;
+        }
```
Device allocations do, so make a new event and record the allocation.

wence- commented on cpp/src/buffer/buffer.cpp:
```diff
@@ -18,24 +23,37 @@ template <typename T>
     RAPIDSMPF_EXPECTS(
         std::get<HostStorageT>(storage_) != nullptr, "the host_buffer cannot be NULL"
     );
     RAPIDSMPF_EXPECTS(br != nullptr, "the BufferResource cannot be NULL");
```
Can't add the check for `event_` being non-null here if the `size` is non-zero because it might not be necessary.

pentschev commented on cpp/src/buffer/buffer.cpp:
```diff
@@ -73,30 +96,55 @@ std::unique_ptr<Buffer> Buffer::copy(MemoryType target, rmm::cuda_stream_view st
-                    ),
-                    br
-                });
+                // Have to ensure that any async work is complete.
+                synchronize();
```
Why?

wence- replied:
This one was wrong, so I fixed it with a stream dependency.

pentschev commented on cpp/src/buffer/buffer.cpp:
```diff
@@ -50,17 +68,22 @@ std::unique_ptr<Buffer> Buffer::copy(rmm::cuda_stream_view stream) const {
-                );
-            },
-            [&](const DeviceStorageT& storage) -> std::unique_ptr<Buffer> {
+                // Have to ensure that any async work is complete.
+                synchronize();
```
Why do you have to ensure async work is complete? Your source is host memory, seems like anything that has been done with it should have been synchronized before, it should never be up to the consumer to magically figure out how the host buffer has been processed, how do you know it is even the right stream then?

wence- replied:
Suppose you spill a device side buffer to `Buffer` as host memory. You did that with `cudaMemcpyAsync`. Now you have a `Buffer` with host memory, but it might not be `ready`. So if you then copy it to a new `Buffer` with host memory you can either fail if it's not `is_ready` or you synchronize.

pentschev replied:
Yeah, ok, but when would that happen in practice? I'd argue the consumer needs to checking the `Buffer` (or whatever other object) that delivered the host memory should be checked if `is_ready()`, otherwise, to use your words, we are now putting a "big hammer" (although slightly smaller than device sync) here. What is the argument for whoever called `Buffer::copy()` not to have ensured that in the first place?

wence- replied:
If we don't do the synchronisation then every time you do a buffer host-to-host copy you must check that that is what you're doing and then `sync`. Or else check `is_ready` and then if you're not ready, push it back into a copy queue?

That seems like a very difficult to manage API.

wence- replied:
In almost all cases the source host buffer will have no event and thus the sync is a noop. But sometimes it might

wence- commented on cpp/src/buffer/buffer.cpp:
```diff
@@ -73,30 +96,55 @@ std::unique_ptr<Buffer> Buffer::copy(MemoryType target, rmm::cuda_stream_view st
-                    ),
-                    br
-                });
+                // Have to ensure that any async work is complete.
+                synchronize();
+                auto buf = std::make_unique<rmm::device_buffer>(
+                    storage->data(), storage->size(), stream, br->device_mr()
+                );
+                event->record(stream);
+
+                return std::unique_ptr<Buffer>(
+                    new Buffer{std::move(buf), br, std::move(event)}
+                );
```
Actually, this one can add a stream dependency as well...

wence- commented:
> and add_as_dependency_to which after a high-level look I still fail to understand how it adds any important information that cannot be capture by recording the event.

Suppose the application stream is `streamA` and the shuffler is internally using a stream `streamB`.

When we take ownership of the application buffer we record the state of `streamA` in an `event`.

Then suppose we spill that data, but we're using `streamB` to call the `cudaMemcpyAsync`. Unless we insert a dependency between `streamA` and `streamB` we will have an invalid result. The way to do this is to call `cudaStreamWaitEvent(streamB, event)` which is what `add_as_dependency_to`.

pentschev commented:
> Suppose the application stream is `streamA` and the shuffler is internally using a stream `streamB`.
> 
> When we take ownership of the application buffer we record the state of `streamA` in an `event`.
> 
> Then suppose we spill that data, but we're using `streamB` to call the `cudaMemcpyAsync`. Unless we insert a dependency between `streamA` and `streamB` we will have an invalid result. The way to do this is to call `cudaStreamWaitEvent(streamB, event)` which is what `add_as_dependency_to` does.

Fair, but to me this is beyond the scope of this change, although since it's already here it's fine, otherwise I would have argued it should be part of a follow-up PR where `Shuffler` uses another stream. I mean we could just do everything we can think of are good changes in one PR, but I feel there's value in keeping things minimal so we don't risk introducing more complexity than necessary and simultaneously risking introducing another bug, but that's just me.

wence- commented:
> > Suppose the application stream is `streamA` and the shuffler is internally using a stream `streamB`.
> > When we take ownership of the application buffer we record the state of `streamA` in an `event`.
> > Then suppose we spill that data, but we're using `streamB` to call the `cudaMemcpyAsync`. Unless we insert a dependency between `streamA` and `streamB` we will have an invalid result. The way to do this is to call `cudaStreamWaitEvent(streamB, event)` which is what `add_as_dependency_to` does.
> 
> Fair, but to me this is beyond the scope of this change, although since it's already here it's fine, otherwise I would have argued it should be part of a follow-up PR where `Shuffler` uses another stream. I mean we could just do everything we can think of are good changes in one PR, but I feel there's value in keeping things minimal so we don't risk introducing more complexity than necessary and simultaneously risking introducing another bug, but that's just me.


I can certainly rip it out, but we still need to think about what to do about host-to-host copies (given that the source buffer in that scenario might have stream-queued work).

wence- commented:
> I can certainly rip it out, but we still need to think about what to do about host-to-host copies (given that the source buffer in that scenario might have stream-queued work).

Added a `RAPIDSMPF_EXPECTS(is_ready())` instead and removed all the `synchronize` APIs and `add_as_dependency_to` calls.

madsbk commented:
Please re-open if this is still relevant.
